binding.cc

```diff
@@ -263,6 +274,246 @@ bare_sdl_get_event_key_scancode(
   return key->handle.scancode;
 }
 
+static uint32_t
+bare_sdl_open_audio_device(
+  js_env_t *env,
+  js_receiver_t,
+  uint32_t requested_device_id,
+  std::optional<uint32_t> format,
+  std::optional<int> channels,
+  std::optional<int> freq
+) {
+  int err;
+
+  SDL_AudioSpec *spec_ptr = nullptr;
+  if (format.has_value() && channels.has_value() && freq.has_value()) {
+    SDL_AudioSpec spec = {
+      .format = (SDL_AudioFormat) format.value(),
+      .channels = channels.value(),
+      .freq = freq.value()
+    };
+
+    spec_ptr = &spec;
+  }
+
+  auto logical_device_id = SDL_OpenAudioDevice((SDL_AudioDeviceID) requested_device_id, spec_ptr);
+
+  if (logical_device_id == 0) {
+    const char *sdl_error = SDL_GetError();
+    err = js_throw_error(env, nullptr, sdl_error);
+    assert(err == 0);
+
+    throw js_pending_exception;
+  }
+
+  return logical_device_id;
+}
+
+static void
+bare_sdl_close_audio_device(
+  js_env_t *,
+  js_receiver_t,
+  uint32_t device_id
+) {
+  SDL_CloseAudioDevice((SDL_AudioDeviceID) device_id);
+}
+
+static bool
+bare_sdl_pause_audio_device(
+  js_env_t *env,
+  js_receiver_t,
+  uint32_t device_id
+) {
+  return SDL_PauseAudioDevice(device_id);
+}
+
+static bool
+bare_sdl_resume_audio_device(
+  js_env_t *env,
+  js_receiver_t,
+  uint32_t device_id
+) {
+  return SDL_ResumeAudioDevice(device_id);
+}
+
+static std::vector<uint32_t>
+bare_sdl_get_audio_playback_devices(
+  js_env_t *env,
+  js_receiver_t
+) {
+  int count = 0;
+  SDL_AudioDeviceID *devices = SDL_GetAudioPlaybackDevices(&count);
+
+  std::vector<uint32_t> list;
+
+  if (devices != nullptr) {
+    for (int i = 0; i < count; i++) {
+      list.push_back(devices[i]);
+    }
+    SDL_free(devices);
+  }
+
+  return list;
+}
+
+static std::vector<uint32_t>
+bare_sdl_get_audio_recording_devices(
+  js_env_t *env,
+  js_receiver_t
+) {
+  int count = 0;
+  SDL_AudioDeviceID *devices = SDL_GetAudioRecordingDevices(&count);
+
+  std::vector<uint32_t> list;
+
+  if (devices != nullptr) {
+    for (int i = 0; i < count; i++) {
+      list.push_back(devices[i]);
+    }
+    SDL_free(devices);
+  }
+
+  return list;
+}
+
+static std::optional<uint32_t>
+bare_sdl_get_audio_device_list_item(
+  js_env_t *env,
+  js_receiver_t,
+  js_arraybuffer_span_of_t<bare_sdl_audio_device_list_t, 1> list,
+  int index
+) {
+  if (index < 0 || index >= list->count || !list->devices) {
+    return NULL;
+  }
+
+  return (uint32_t) list->devices[index];
+}
+
+static std::string
+bare_sdl_get_audio_device_name(
+  js_env_t *env,
+  js_receiver_t,
+  uint32_t device_id
+) {
+  const char *name = SDL_GetAudioDeviceName((SDL_AudioDeviceID) device_id);
+  return name ? std::string(name) : std::string("");
```

Comment:
```suggestion
  return name ? std::string(name) : "";
```

It does beg the question if the return value should be `std:optional<std::string>`.

Reply:
I went ahead and made this return `std::optional<std::string>`. Feels right.